README.md
# ![logo](media/EMS-ESP_logo_dark.png)

**EMS-ESP** is an open-source firmware for the Espressif ESP8266 and ESP32 microcontroller that communicates with **EMS** (Energy Management System) based equipment from manufacturers like Bosch, Buderus, Nefit, Junkers, Worcester and Sieger.

This is the firmware for the ESP8266, known as version 2. We've since moved to the ESP32 platform with v3 which opens up many more possibilities for additional features. Make sure you check out <https://github.com/emsesp/EMS-ESP32> and look at the [demo](https://ems-esp.derbyshire.nl).

## **Important: This version will be restricted to maintenance releases only!**
<br/>

[![version](https://img.shields.io/github/release/emsesp/EMS-ESP.svg?label=Latest%20Release)](https://github.com/emsesp/EMS-ESP/blob/main/CHANGELOG.md)
[![release-date](https://img.shields.io/github/release-date/emsesp/EMS-ESP.svg?label=Released)](https://github.com/emsesp/EMS-ESP/commits/main)
[![license](https://img.shields.io/github/license/emsesp/EMS-ESP.svg)](LICENSE)
[![downloads](https://img.shields.io/github/downloads/emsesp/EMS-ESP/total.svg)](https://github.com/emsesp/EMS-ESP/releases)
[![Average time to resolve an issue](http://isitmaintained.com/badge/resolution/emsesp/EMS-ESP.svg)](http://isitmaintained.com/project/emsesp/EMS-ESP "Average time to resolve an issue")
[![Percentage of issues still open](http://isitmaintained.com/badge/open/emsesp/EMS-ESP.svg)](http://isitmaintained.com/project/emsesp/EMS-ESP "Percentage of issues still open")
<br/>
[![chat](https://img.shields.io/discord/816637840644505620.svg?style=flat-square&color=blueviolet)](https://discord.gg/3J3GgnzpyT)

If you like **EMS-ESP**, please give it a star, or fork it and contribute!

[![GitHub stars](https://img.shields.io/github/stars/emsesp/EMS-ESP.svg?style=social&label=Star)](https://github.com/emsesp/EMS-ESP/stargazers)
[![GitHub forks](https://img.shields.io/github/forks/emsesp/EMS-ESP.svg?style=social&label=Fork)](https://github.com/emsesp/EMS-ES32P/network)
[![donate](https://img.shields.io/badge/donate-PayPal-blue.svg)](https://www.paypal.com/paypalme/prderbyshire/2)

Note, EMS-ESP requires a small hardware circuit that can convert the EMS bus data to be read by the microcontroller. These can be ordered at <https://bbqkees-electronics.nl>.

<img src="media/gateway-integration.jpg" width=40%>

---

## **Features**

- Compatible with both ESP8266 and ESP32
- A multi-user secure web interface to change settings and monitor the data
- A console, accessible via Serial and Telnet for more monitoring
- Native support for Home Assistant via [MQTT Discovery](https://www.home-assistant.io/docs/mqtt/discovery/)
- Can run standalone as an independent WiFi Access Point or join an existing WiFi network
- Easy first-time configuration via a web Captive Portal
- Support for more than [70 EMS devices](https://emsesp.github.io/docs/#/Supported-EMS-Devices) (boilers, thermostats, solar modules, mixer modules, heat pumps, gateways)
  
## **Screenshots**

### Web Interface

| | |
| --- | --- |
| <img src="media/web_settings.PNG"> | <img src="media/web_status.PNG"> |
| <img src="media/web_devices.PNG"> | <img src="media/web_mqtt.PNG"> |

### Telnet Console

<img src="media/console.PNG" width=80% height=80%>

### In Home Assistant

<img src="media/ha_lovelace.PNG" width=80% height=80%>
  
## **Installing**

Refer to the [official documentation](https://emsesp.github.io/docs) to how to install the firmware and configure it. The documentation is being constantly updated as new features and settings are added.

You can choose to use an pre-built firmware image or compile the code yourself:

- [Uploading a pre-built firmware build](https://emsesp.github.io/docs/#/Uploading-firmware)
- [Building the firmware from source code and flashing manually](https://emsesp.github.io/docs/#/Building-firmware)

## **Support Information**

If you're looking for support on **EMS-ESP** there are some options available:

### Documentation

- [Official EMS-ESP Documentation](https://emsesp.github.io/docs): For information on how to build and upload the firmware
- [FAQ and Troubleshooting](https://emsesp.github.io/docs/#/Troubleshooting): For information on common problems and solutions. See also [BBQKees's wiki](https://bbqkees-electronics.nl/wiki/gateway/troubleshooting.html)

### Support Community

- [Discord Server](https://discord.gg/3J3GgnzpyT): For support, troubleshooting and general questions. You have better chances to get fast answers from members of the community
- [Search in Issues](https://github.com/emsesp/EMS-ESP/issues): You might find an answer to your question by searching current or closed issues

### Developer's Community

- [Bug Report](https://github.com/emsesp/EMS-ESP/issues/new?template=bug_report.md): For reporting Bugs
- [Feature Request](https://github.com/emsesp/EMS-ESP/issues/new?template=feature_request.md): For requesting features/functions
- [Troubleshooting](https://github.com/emsesp/EMS-ESP/issues/new?template=questions---troubleshooting.md): As a last resort, you can open new *Troubleshooting & Question* issue on GitHub if the solution could not be found using the other channels. Just remember: the more info you provide the more chances you'll have to get an accurate answer

# **Contributors ✨**

EMS-ESP is a project originally created by [proddy](https://github.com/proddy) with the main contributors and owners:

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore-start -->
<!-- markdownlint-disable -->
<table>
  <tr>
    <td align="center">
    <a href="https://derbyshire.nl"><img src="https://avatars.githubusercontent.com/u/1230712?v=3?s=100" width="100px;" alt=""/>
    <br/><sub><b>proddy</b></sub></a>
    <br/>
    </a> <a href="https://github.com/emsesp/EMS-ESP/commits?author=proddy" title="v2 Commits">v2</a>
    <a href="https://github.com/emsesp/EMS-ESP32/commits?author=proddy" title="v3 Commits">v3</a>
    </td>
    <td align="center">
    <a href="https://github.com/MichaelDvP"><img src="https://avatars.githubusercontent.com/u/59284019?v=3?s=100" width="100px;" alt=""/><br /><sub><b>MichaelDvP</b></sub></a><br /></a> <a href="https://github.com/emsesp/EMS-ESP/commits?author=MichaelDvP" title="v2 Commits">v2</a>
    <a href="https://github.com/emsesp/EMS-ESP32/commits?author=MichaelDvP" title="v3 Commits">v3</a>
    </td>

  </tr>
</table>
<!-- markdownlint-restore -->
<!-- prettier-ignore-end -->

You can also contribute to EMS-ESP by

- providing Pull Requests (Features, Fixes, suggestions)
- testing new released features and report issues on your EMS equipment
- contributing to missing [Documentation](https://emsesp.github.io/docs)

# **Libraries used**

- [esp8266-react](https://github.com/rjwats/esp8266-react) by @rjwats for the framework that provides the  Web UI
- [uuid-\*](https://github.com/nomis/mcu-uuid-console) from @nomis. The console, syslog, telnet and logging is based on these libraries
- [ArduinoJson](https://github.com/bblanchon/ArduinoJson)
- [AsyncMqttClient](https://github.com/marvinroger/async-mqtt-client) for MQTT, with modifications from @bertmelis
- ESPAsyncWebServer and ESPAsyncTCP for the Web and TCP backends, with custom modifications for performance

# **License**

This program is licensed under GPL-3.0


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.2.2] May 4 2021

## Added

- differ between boiler-types
- publish HA delayed 1 minute, only for topics with values
- thermostat CRF200S, PID 216

## Changed

- syslog timestamp to localtime+timezone, set `appname` to message->name

## Removed

## [2.2.1] March 18 2021

### Added

- Boiler WB Greenstar 8000 (195) and Cascade Modul MC400 (210)
- show cascaded boilers
- optional id to info command to output heatingcircuits separated
- Reset to factory setting with PButton (io0 to GND for >9 sec)

### Fixed

- telegrams matched to masterthermostat 0x18
- Boiler Junkers Cerapur Aero
- multiple roomcontrollers
- readback after write with delay (give ems-devices time to set the value)
- Thermostat ES72/RC20, device 66 to RC20_2 command-set
- recognize sending devices which are not in telegram 0x07
- MQTT payloads not adding to queue when MQTT is re-connecting (fixes #369)
- fix mixerTemp and tankMiddleTemp (PR #714 @joanwa)
- icon update for pumps
- add mixer wiring naming from documentation
- wwtemp feedback and ems+ command
- fix for HA topics with invalid command formats - #728
- fix ems+ values #723, #732

### Changed

- split `show values` in smaller packages and separate heating circuits
- extended length of IP/hostname from 32 to 48 chars (#676)
- check flowsensor for `tap_water_active`
- serviceCode `~H` instead of non-ascii 3-dashes
- mqtt `base` instead of hostname
- mqtt don't store full topics, add base on publish to save memory
- mqtt max. topic length fixed to 128
- thermostat `time` to `datetime`
- adc scaled to mV output
- split boiler mqtt to `data`, `data_ww`, `data_info`
- Reduce Web UI artefact size by removing moment.js

## [2.2.0] December 28 2020

### Added

- function keys in editor: cursor, del, home, end. F1=help, F2=show, and other shortcuts
- SM100 pump working time and energy units
- heating curve parameters and commands for RC300
- `wwonetime` for RC300 thermostat
- expose test framework via api (#611)
- SysLog has enable/disable flag in WebUI
- Add solar configuration telegrams (#616) [thanks @hpanther]
- `log trace` shows decoded or optional raw telegrams, `watch unknown` for only unknown telegrams
- WM10 switch telegrams
- boiler information (#633), pumpmod min/max commands
- maintenance message and command
- thermostat program, reducemode, controlmode
- optional delayed start for sending tx-telegrams to prevent conflicts with KM200
- RC35 holiday setting with `+` for 'at home'

### Fixed

- mixer IPM pumpstatus
- mixer devices in HA were incorrectly named
- Prevent HA MQTT config messages for thermostat that has no 'currtemp' (#582)
- serviceCodeNumber, 3-char serviceCode, exhausttemp and heating_active for newer ems+ boilers
- prevent MQTT publish messages from sending twice
- repeated output on read commands
- heating_active for ems+
- fix telegrams matched to masterthermostat 0x18

### Changed

- optimized MQTT for HA to reduce heap fragmentation issues
- change syslog settings without reboot
- HA-config split in smaller blocks
- commands `fetch` and `publish [ha]` as call
- mqtt json package sizes
- renamed the command system info (which showed settings) to `settings`
- renamed the command system report (Which dumped debug info) to `info`
- Changing settings via web restarts only selected services
- renamed pio targets (esp8266-ci and esp32-ci for GitHub CI)
- telnet default settings `log info`, timeout 60 min
- `log debug` not showing telegram names, use `log trace` or `watch on` to show the telegrams
- optimized how console and web display device data ([#632](https://github.com/emsesp/EMS-ESP/issues/632))

### Removed

- old shell and python build scripts

## [2.1.0] October 31 2020

### Added

- boiler `heatingactivated`, automatic select parameter telegrams for write
- boiler `wWType` parameter, in Console and MQTT
- support for uploading compressed firmware binaries in web UI
- setting to manually override the MQTT retain flag
- New API via HTTP REST API to read and set values. See <https://emsesp.github.io/docs/#/API>
- `show commands` command
- exporting of system settings using the `system info` command in Web and Console. Added link into the Web's Settings page.
- setting to change how booleans are rendered in MQTT (on/off, true/false, 1/0)
- enable ADC setting, add boiler circulation commands, add thermostat RC300 summermodes
- Added all device info to web UI for Thermostat and Boiler
- Added all device values to Home Assistant MQTT Discovery under separate devices and entities
- Show Rx and Tx quality in Console and Web UI
- Added button and tooltip to EMS Devices in Web
- wwtemp and wwtemplow to MQTT, Console and Web
- summer, winter modes for the CW400 thermostat
- new command under system called `report`. <http://ems-esp/api?device=system&cmd=report> to generate a report log for troubleshooting
- thermostat error codes
- Console command `publish ha` to also force the creation of the Home Assistant MQTT Discovery topics
- Heat pump values (dew temperature and relative air humidity)
- Console up key to repeat last command
- added RC300 floordrying, building, damped temperature

### Fixed

- fix wwontime readback
- fixed support for RC300 via MQTT commands (#505)
- Some minor optimizations to memory handling in the MQTT service

src/MyESP.h
/*
 * MyESP.h - does all the basics like WiFI/MQTT/NTP/Debug logs etc
 *
 * Paul Derbyshire - first version December 2018
 */

#pragma once

#ifndef MyESP_h
#define MyESP_h

#define MYESP_VERSION "1.2.30"

#include <ArduinoJson.h>
#include <ArduinoOTA.h>
#include <AsyncMqttClient.h>
#include <ESPAsyncWebServer.h>
#include <FS.h>
#include <JustWifi.h>
#include <deque> // for MQTT publish queue

// SysLog
#include <uuid/common.h>
#include <uuid/log.h>
#include <uuid/syslog.h>
static uuid::syslog::SyslogService syslog;
enum MYESP_SYSLOG_LEVEL : uint8_t { MYESP_SYSLOG_INFO, MYESP_SYSLOG_ERROR };

// local libraries
#include "Ntp.h"
#include "TelnetSpy.h" // modified from https://github.com/yasheena/telnetspy

#ifdef CRASH
#include <EEPROM_Rotate.h>
#endif

extern "C" {
void custom_crash_callback(struct rst_info *, uint32_t, uint32_t);
#include "user_interface.h"
extern struct rst_info resetInfo;
}

#if defined(ARDUINO_ARCH_ESP32)
#include <SPIFFS.h>             // added for ESP32
#define ets_vsnprintf vsnprintf // added for ESP32
#define OTA_PORT 3232
#else
#include <ESPAsyncTCP.h>
#define OTA_PORT 8266
#endif

// web files
// reference libs
#include "webh/glyphicons-halflings-regular.woff.gz.h"
#include "webh/required.css.gz.h"
#include "webh/required.js.gz.h"

// custom stuff
#include "webh/index.html.gz.h"
#include "webh/myesp.html.gz.h"
#include "webh/myesp.js.gz.h"

#define MYESP_CONFIG_FILE "/myesp.json"
#define MYESP_CUSTOMCONFIG_FILE "/customconfig.json"
#define MYESP_OLD_EVENTLOG_FILE "/eventlog.json" // depreciated
#define MYESP_OLD_CONFIG_FILE "/config.json"     // depreciated

#define MYESP_HTTP_USERNAME "admin" // HTTP username
#define MYESP_HTTP_PASSWORD "admin" // default password

#define MYESP_NTP_SERVER "pool.ntp.org" // default ntp server

#define MYESP_LOADAVG_INTERVAL 30000 // Interval between calculating load average (in ms) = 30 seconds

// WIFI
#define MYESP_WIFI_CONNECT_TIMEOUT 20000     // Connecting timeout for WIFI in ms (20 seconds)
#define MYESP_WIFI_RECONNECT_INTERVAL 600000 // If could not connect to WIFI, retry after this time in ms. 10 minutes

// set to value >0 if the ESP is overheating or there are timing issues. Recommend a value of 1.
#define MYESP_DELAY 1

// MQTT
#define MQTT_PORT 1883                  // MQTT port
#define MQTT_RECONNECT_DELAY_MIN 2000   // Try to reconnect in 3 seconds upon disconnection
#define MQTT_RECONNECT_DELAY_STEP 3000  // Increase the reconnect delay in 3 seconds after each failed attempt
#define MQTT_RECONNECT_DELAY_MAX 120000 // Set reconnect time to 2 minutes at most
#define MQTT_TOPIC_START "start"
#define MQTT_TOPIC_HEARTBEAT "heartbeat"
#define MQTT_TOPIC_START_PAYLOAD "start"
#define MQTT_TOPIC_RESTART "restart"
#define MQTT_WILL_ONLINE_PAYLOAD "online"   // for last will & testament payload
#define MQTT_WILL_OFFLINE_PAYLOAD "offline" // for last will & testament payload
#define MQTT_BASE_DEFAULT "home"            // default MQTT prefix to topics
#define MQTT_RETAIN false                   // default false
#define MQTT_KEEPALIVE 60                   // default keepalive 1 minute
#define MQTT_QOS 0                          // default qos 0
#define MQTT_WILL_TOPIC "status"            // for last will & testament topic name
#define MQTT_MAX_TOPIC_SIZE 50              // max length of MQTT topic
#define MQTT_QUEUE_MAX_SIZE 50              // Size of the MQTT queue
#define MQTT_PUBLISH_WAIT 750               // time in ms before sending MQTT messages
#define MQTT_PUBLISH_MAX_RETRY 4            // max retries for giving up on publishing
#define MYESP_JSON_MAXSIZE_LARGE 2000       // for large Dynamic json files - https://arduinojson.org/v6/assistant/
#define MYESP_JSON_MAXSIZE_MEDIUM 800       // for medium Dynamic json files - https://arduinojson.org/v6/assistant/
#define MYESP_JSON_MAXSIZE_SMALL 200        // for smaller Static json documents - https://arduinojson.org/v6/assistant/

// Internal MQTT events
#define MQTT_CONNECT_EVENT 0
#define MQTT_DISCONNECT_EVENT 1
#define MQTT_MESSAGE_EVENT 2

#define MYESP_MQTT_PAYLOAD_ON '1'  // for MQTT switch on
#define MYESP_MQTT_PAYLOAD_OFF '0' // for MQTT switch off

// Telnet
#define TELNET_SERIAL_BAUD 115200
#define TELNET_MAX_COMMAND_LENGTH 80 // length of a command
#define TELNET_MAX_BUFFER_LENGTH 700 // max length of telnet string
#define TELNET_EVENT_CONNECT 1
#define TELNET_EVENT_DISCONNECT 0
#define TELNET_EVENT_SHOWCMD 10
#define TELNET_EVENT_SHOWSET 20

// ANSI Colors
#define COLOR_RESET "\x1B[0m"
#define COLOR_BLACK "\x1B[0;30m"
#define COLOR_RED "\x1B[0;31m"
#define COLOR_GREEN "\x1B[0;32m"
#define COLOR_YELLOW "\x1B[0;33m"
#define COLOR_BLUE "\x1B[0;34m"
#define COLOR_MAGENTA "\x1B[0;35m"
#define COLOR_CYAN "\x1B[0;36m"
#define COLOR_WHITE "\x1B[0;37m"
#define COLOR_BOLD_ON "\x1B[1m"
#define COLOR_BOLD_OFF "\x1B[22m"
#define COLOR_BRIGHT_BLACK "\x1B[0;90m"
#define COLOR_BRIGHT_RED "\x1B[0;91m"
#define COLOR_BRIGHT_GREEN "\x1B[0;92m"
#define COLOR_BRIGHT_YELLOW "\x1B[0;99m"
#define COLOR_BRIGHT_BLUE "\x1B[0;94m"
#define COLOR_BRIGHT_MAGENTA "\x1B[0;95m"
#define COLOR_BRIGHT_CYAN "\x1B[0;96m"
#define COLOR_BRIGHT_WHITE "\x1B[0;97m"

// reset reason codes
PROGMEM const char         custom_reset_hardware[] = "Hardware button";
PROGMEM const char         custom_reset_terminal[] = "Restart from terminal";
PROGMEM const char         custom_reset_mqtt[]     = "Restart from MQTT";
PROGMEM const char         custom_reset_ota[]      = "Restart after successful OTA update";
PROGMEM const char         custom_reset_factory[]  = "Factory reset";
PROGMEM const char * const custom_reset_string[]   = {custom_reset_hardware, custom_reset_terminal, custom_reset_mqtt, custom_reset_ota, custom_reset_factory};
#define CUSTOM_RESET_HARDWARE 1 // Reset from hardware button
#define CUSTOM_RESET_TERMINAL 2 // Reset from terminal
#define CUSTOM_RESET_MQTT 3     // Reset via MQTT
#define CUSTOM_RESET_OTA 4      // Reset after successful OTA update
#define CUSTOM_RESET_FACTORY 5  // Factory reset
#define CUSTOM_RESET_MAX 5

// SPIFFS - max allocation is 1000 KB
// https://arduinojson.org/v6/assistant/
#define MYESP_SPIFFS_MAXSIZE_CONFIG 999 // max size for a config file

// CRASH
/**
 * Structure of the single crash data set
 *
 *  1. Crash time
 *  2. Restart reason
 *  3. Exception cause
 *  4. epc1
 *  5. epc2
 *  6. epc3
 *  7. excvaddr
 *  8. depc
 *  9. address of stack start
 * 10. address of stack end
 * 11. stack trace bytes
 *     ...
 */
#define SAVE_CRASH_EEPROM_OFFSET 0x0100 // initial address for crash data
#define SAVE_CRASH_CRASH_TIME 0x00      // 4 bytes
#define SAVE_CRASH_RESTART_REASON 0x04  // 1 byte
#define SAVE_CRASH_EXCEPTION_CAUSE 0x05 // 1 byte
#define SAVE_CRASH_EPC1 0x06            // 4 bytes
#define SAVE_CRASH_EPC2 0x0A            // 4 bytes
#define SAVE_CRASH_EPC3 0x0E            // 4 bytes
#define SAVE_CRASH_EXCVADDR 0x12        // 4 bytes
#define SAVE_CRASH_DEPC 0x16            // 4 bytes
#define SAVE_CRASH_STACK_START 0x1A     // 4 bytes
#define SAVE_CRASH_STACK_END 0x1E       // 4 bytes
#define SAVE_CRASH_STACK_TRACE 0x22     // variable

// Base address of USER RTC memory
// https://github.com/esp8266/esp8266-wiki/wiki/Memory-Map#memmory-mapped-io-registers
#define RTCMEM_ADDR_BASE (0x60001200)

// RTC memory is accessed using blocks of 4 bytes.
// Blocks 0..63 are reserved by the SDK, 64..192 are available to the user.
// Blocks 64..96 are reserved by the eboot 'struct eboot_command' (128 -> (128 / 4) -> 32):
// https://github.com/esp8266/Arduino/blob/master/bootloaders/eboot/eboot_command.h
#define RTCMEM_OFFSET 32u
#define RTCMEM_ADDR (RTCMEM_ADDR_BASE + (RTCMEM_OFFSET * 4u))
#define RTCMEM_BLOCKS 96u
#define RTCMEM_MAGIC 0x45535076

struct RtcmemData {
    uint32_t magic; // RTCMEM_MAGIC
    uint32_t sys;   // system details
};

static_assert(sizeof(RtcmemData) <= (RTCMEM_BLOCKS * 4u), "RTCMEM struct is too big");

#define MYESP_SYSTEM_CHECK_TIME 60000  // The system is considered stable after these many millis (1 min)
#define MYESP_SYSTEM_CHECK_MAX 10      // After this many crashes on boot
#define MYESP_HEARTBEAT_INTERVAL 60000 // in milliseconds, how often the MQTT heartbeat is sent (1 min)

typedef struct {
    bool set; // is it a set command?
    char key[50];
    char description[110];
} command_t;

typedef enum {
    MYESP_FSACTION_SET,
    MYESP_FSACTION_LIST,
    MYESP_FSACTION_SAVE,
    MYESP_FSACTION_LOAD,
    MYESP_FSACTION_ERR,
    MYESP_FSACTION_OK,
    MYESP_FSACTION_RESTART
} MYESP_FSACTION_t;

typedef enum {
    MYESP_BOOTSTATUS_POWERON     = 0,
    MYESP_BOOTSTATUS_BOOTED      = 1,
    MYESP_BOOTSTATUS_BOOTING     = 2,
    MYESP_BOOTSTATUS_RESETNEEDED = 3
} MYESP_BOOTSTATUS_t; // boot messages

typedef std::function<void(unsigned int, const char *, const char *)>                          mqtt_callback_f;
typedef std::function<void()>                                                                  wifi_callback_f;
typedef std::function<void()>                                                                  ota_callback_f;
typedef std::function<void(char *commandLine, int argc, char **argv)>                          telnetcommand_callback_f;
typedef std::function<void(uint8_t)>                                                           telnet_callback_f;
typedef std::function<bool(MYESP_FSACTION_t, JsonObject json)>                                 fs_loadsave_callback_f;
typedef std::function<MYESP_FSACTION_t(MYESP_FSACTION_t, char **, size_t)> fs_setlist_callback_f;
typedef std::function<void(JsonObject root)>                                                   web_callback_f;

// calculates size of an 2d array at compile time
template <typename T, size_t N>
constexpr size_t ArraySize(T (&)[N]) {
    return N;
}

#define MYESP_UPTIME_OVERFLOW 4294967295 // Uptime overflow value
#define MYESP_MAX_STR_LEN 16             // web min and max length of wifi ssid and password
#define MYESP_BOOTUP_FLASHDELAY 50       // flash duration for LED at bootup sequence
#define MYESP_BOOTUP_DELAY 2000          // time before we open the window to reset. This is to stop resetting values when uploading firmware via USB

// class definition
class MyESP {
  protected:
    // webserver
    AsyncWebServer * _webServer;
    AsyncWebSocket * _ws;

    // NTP
    NtpClient NTP;

  public:
    MyESP();
    ~MyESP();

    // wifi
    void setWIFICallback(void (*callback)());
    void setWIFI(wifi_callback_f callback);
    bool isWifiConnected();
    bool isAPmode();

    // mqtt
    bool isMQTTConnected();
    bool isMQTTHealthy();
    bool mqttSubscribe(const char * topic);
    void mqttUnsubscribe(const char * topic);
    void mqttPublish(const char * topic, const char * payload);
    void mqttPublish(const char * topic, const char * payload, bool retain);
    void mqttPublish(const char * topic, JsonDocument payload);
    void mqttPublish(const char * topic, JsonDocument payload, bool retain);
    void setMQTT(mqtt_callback_f callback);
    bool mqttUseNestedJson();

    // OTA
    void setOTA(ota_callback_f OTACallback_pre, ota_callback_f OTACallback_post);

    // debug & telnet
    void myDebug(const char * format, ...);
    void myDebug_P(PGM_P format_P, ...);
    void setTelnet(telnetcommand_callback_f callback_cmd, telnet_callback_f callback);
    bool getUseSerial();
    void setUseSerial(bool toggle);

    // syslog
    void writeLogEvent(const uint8_t type, const char * msg);

    // FS
    void setSettings(fs_loadsave_callback_f loadsave, fs_setlist_callback_f setlist, bool useSerial = true);
    void saveSettings();
    bool fs_saveConfig(JsonObject root);
    bool fs_saveCustomConfig(JsonObject root);
    bool fs_setSettingValue(char ** setting, const char * value, const char * value_default);
    bool fs_setSettingValue(uint16_t * setting, const char * value, uint16_t value_default);
    bool fs_setSettingValue(uint8_t * setting, const char * value, uint8_t value_default);
    bool fs_setSettingValue(int8_t * setting, const char * value, int8_t value_default);
    bool fs_setSettingValue(bool * setting, const char * value, bool value_default);

    // Web
    void setWeb(web_callback_f callback_web);

    // Crash
    void crashClear();
    void crashDump();
    void crashInfo();

    // general
    void          end();
    void          loop();
    void          begin(const char * app_hostname, const char * app_name, const char * app_version, const char * app_url, const char * app_url_api);
    void          resetESP();
    int           getWifiQuality();
    void          showSystemStats();
    bool          getHeartbeat();
    uint32_t      getSystemLoadAverage();
    uint32_t      getSystemResetReason();
    uint8_t       getSystemBootStatus();
    unsigned long getSystemTime();
    void          heartbeatPrint();
    void          heartbeatCheck(bool force = false);

  private:
    // mqtt
    void            _mqttOnMessage(char * topic, char * payload, size_t len);
    void            _mqttOnPublish(uint16_t packetId);
    void            _mqttConnect();
    void            _mqtt_setup();
    void            _mqttOnConnect();
    void            _sendStart();
    char *          _mqttTopic(const char * topic);
    bool            _mqttQueue(const char * topic, const char * payload, bool retain);
    bool            _mqttQueue(const char * topic, JsonDocument payload, bool retain);
    void            _printMQTTLog();
    void            _mqttPublishQueue();
    void            _mqttRemoveLastPublish();
    AsyncMqttClient mqttClient; // the MQTT class
    uint32_t        _mqtt_reconnect_delay;
    mqtt_callback_f _mqtt_callback_f;
    char *          _mqtt_ip;
    char *          _mqtt_user;
    char *          _mqtt_password;
    uint16_t        _mqtt_port;
    char *          _mqtt_base;
    bool            _mqtt_enabled;
    uint16_t        _mqtt_keepalive;
    uint8_t         _mqtt_qos;
    bool            _mqtt_retain;
    char *          _mqtt_will_topic;
    char *          _mqtt_will_online_payload;
    char *          _mqtt_will_offline_payload;
    uint32_t        _mqtt_last_connection;
    bool            _mqtt_connecting;
    bool            _mqtt_heartbeat;
    bool            _mqtt_nestedjson;
    uint16_t        _mqtt_publish_fails;
    uint8_t         _mqtt_publish_failed;

    // wifi
    void            _wifiCallback(justwifi_messages_t code, char * parameter);
    void            _wifi_setup();
    wifi_callback_f _wifi_callback_f;
    char *          _network_ssid;
    char *          _network_password;
    uint8_t         _network_wmode;
    char *          _network_staticip;
    char *          _network_gatewayip;
    char *          _network_nmask;
    char *          _network_dnsip;
    bool            _wifi_connected;
    String          _getESPhostname();

    // ota
    ota_callback_f _ota_pre_callback_f;
    ota_callback_f _ota_post_callback_f;
    void           _ota_setup();
    void           _OTACallback();

    // crash
    void _eeprom_setup();

    // telnet
    TelnetSpy                SerialAndTelnet;
    void                     _telnetConnected();
    void                     _telnetDisconnected();
    void                     _telnetHandle();
    void                     _telnetCommand(char * commandLine, int argc, char **argv);
    void                     _telnetCommandParser(char * commandLine);
    //char *                   _telnet_readWord(bool allow_all_chars);
    void                     _telnet_setup();
    char                     _command[TELNET_MAX_COMMAND_LENGTH + 1]; // the input command from either Serial or Telnet
    void                     _consoleShowHelp();
    telnetcommand_callback_f _telnetcommand_callback_f; // Callable for projects commands
    telnet_callback_f        _telnet_callback_f;        // callback for connect/disconnect
    bool                     _changeSetting(int argc, char **argv);

    // syslog
    void _syslog_setup();

    // fs and settings
    void                   _fs_setup();
    bool                   _fs_loadConfig();
    bool                   _fs_loadCustomConfig();
    void                   _fs_eraseConfig();
    bool                   _fs_writeConfig();
    bool                   _fs_createCustomConfig();
    bool                   _fs_sendConfig();
    size_t                 _fs_validateConfigFile(const char * filename, size_t maxsize, JsonDocument & doc);
    size_t                 _fs_validateLogFile(const char * filename);
    fs_loadsave_callback_f _fs_loadsave_callback_f;
    fs_setlist_callback_f  _fs_setlist_callback_f;
    void                   _printSetCommands();

    // general
    char *        _general_hostname;
    char *        _app_name;
    char *        _app_version;
    char *        _app_url;
    char *        _app_updateurl;
    char *        _app_updateurl_dev;
    bool          _suspendOutput;
    bool          _general_serial;
    bool          _general_log_events;
    char *        _general_log_ip;
    char *        _buildTime;
    bool          _timerequest;
    bool          _formatreq;
    unsigned long _getUptime();
    char *        _getBuildTime();
    bool          _hasValue(const char * s);
    void          _printHeap(const char * s);
    void          _kick();

    // reset reason and rtcmem
    bool    _rtcmem_status;
    bool    _rtcmemStatus();
    bool    _getRtcmemStatus();
    void    _rtcmemInit();
    void    _rtcmemSetup();
    void    _deferredReset(unsigned long delay, uint8_t reason);
    uint8_t _getSystemStabilityCounter();
    void    _setSystemStabilityCounter(uint8_t counter);
    uint8_t _getSystemDropoutCounter();
    void    _setSystemDropoutCounter(uint8_t counter);
    void    _increaseSystemDropoutCounter();
    void    _setSystemResetReason(uint8_t reason);
    uint8_t _getCustomResetReason();
    void    _setCustomResetReason(uint8_t reason);
    uint8_t _getSystemResetReason();
    void    _setSystemBootStatus(uint8_t status);
    bool    _systemStable;
    void    _bootupSequence();
    bool    _getSystemCheck();
    void    _systemCheckLoop();
    void    _setSystemCheck(bool stable);

    // load average (0..100) and heap ram
    void     _calculateLoad();
    uint32_t _load_average;
    uint32_t _getInitialFreeHeap();
    uint32_t _getUsedHeap();

    // web
    web_callback_f _web_callback_f;
    const char *   _http_username;

    // web
    void _onWsEvent(AsyncWebSocket * server, AsyncWebSocketClient * client, AwsEventType type, void * arg, uint8_t * data, size_t len);
    void _procMsg(AsyncWebSocketClient * client, size_t sz);
    void _sendStatus();
    void _sendCustomStatus();
    void _printScanResult(int networksFound);
    void _sendTime();
    void _webserver_setup();

    // ntp
    char *   _ntp_server;
    uint16_t _ntp_interval;
    bool     _ntp_enabled;
    uint8_t  _ntp_timezone;
    bool     _have_ntp_time;
};

extern MyESP myESP;

#endif
